jdaviz/utils.py

```diff
@@ -385,3 +390,72 @@ def total_masked_first_data(self):
     def __setgluestate__(cls, rec, context):
         masks = {key: context.object(value) for key, value in rec['masks'].items()}
         return cls(masks=masks)
+
+
+def download_uri_to_path(possible_uri, cache=False, local_path=None):
+    """
+    Retrieve data from a URI (or a URL). Return the input if it
+    cannot be parsed as a URI.
+
+    If ``possible_uri`` is a MAST URI, the file will be retrieved via
+    astroquery's `~astroquery.mast.ObservationsClass.download_file`.
+    If ``possible_uri`` is a URL, it will be retrieved via astropy with
+    `~astropy.utils.data.download_file`.
+
+    Parameters
+    ----------
+    possible_uri : str or other
+        This input will be returned without changes if it is not a string,
+        or if it is a local file path to an existing file. Otherwise,
+        it will be parsed as a URI. Local file URIs beginning with ``file://``
+        are not supported by this method – nor are they necessary, since string
+        paths without the scheme work fine! Cloud FITS are not yet supported.
+    cache: bool or "update", optional
+        Cache file after download. Default is False. If ``possible_uri`` is a
+        URL, ``cache`` may be a boolean or ``"update"``, see documentation for
+        `~astropy.utils.data.download_file` for details.
+    local_path : str, optional
+        Save the downloaded file to this path. Default is to
+        save the file with its remote filename in the current
+        working directory.
+
+    Returns
+    -------
+    possible_uri : str or other
+        If ``possible_uri`` cannot be retrieved as a URI, returns the input argument
+        unchanged. If ``possible_uri`` can be retrieved as a URI, returns the
+        local path to the downloaded file.
+    """
+    if not isinstance(possible_uri, str):
+        # only try to parse strings:
+        return possible_uri
+
+    if os.path.exists(possible_uri):
+        # don't try to parse file paths:
+        return possible_uri
+
+    parsed_uri = urlparse(possible_uri)
+
+    if parsed_uri.scheme.lower() == 'mast':
+        Observations.download_file(possible_uri, cache=cache, local_path=local_path)
+
+        if local_path is None:
+            # if not specified, this is the default location:
+            local_path = os.path.join(os.getcwd(), parsed_uri.path.split('/')[-1])
+
+        return local_path
+
+    elif parsed_uri.scheme.lower() in ['http', 'https', 'ftp']:
+        return download_file(possible_uri, cache=cache)
+
+    elif parsed_uri.scheme == '':
+        raise ValueError(f"The data to load '{possible_uri}' cannot be parse as a "
+                         f"URL or URI, and no existing local file is available "
+                         f"at this path.")
```

Comment:
```suggestion
        raise ValueError(f"The input file '{possible_uri}' cannot be parsed as a "
                         f"URL or URI, and no existing local file is available "
                         f"at this path.")
```

parse -> parsed and another slight change.